src/client/tests/SessionPoolTest.cpp

```diff
@@ -0,0 +1,288 @@
+
+  std::vector<std::thread> threads;
+  for (std::size_t i = 0; i < config.maxSize_; ++i) {
+    threads.emplace_back([&pool]() {
+      // using namespace std::chrono_literals;  // NOLINT
+      // std::this_thread::sleep_for(1s);
+
+      auto resp = pool.execute("YIELD 1");
+      ASSERT_EQ(resp.errorCode, nebula::ErrorCode::SUCCEEDED) << *resp.errorMsg;
+      nebula::DataSet expected({"1"});
+      expected.emplace_back(nebula::Row({1}));
+      ASSERT_TRUE(verifyResultWithoutOrder(resp, expected));
+    });
+  }
+  for (auto& t : threads) {
+    t.join();
+  }
+}
+
+// case 1
+TEST_F(SessionPoolTest, Exception) {
+  // space don't exists
+  {
+    nebula::SessionPoolConfig config;
+    config.username_ = "root";
+    config.password_ = "nebula";
+    config.addrs_ = {kServerHost ":9669"};
+    config.spaceName_ = "not_exists_space";
+    config.maxSize_ = 2;
+    nebula::SessionPool pool(config);
+    ASSERT_FALSE(pool.init());
+  }
+
+  // user don't exists
+  {
+    nebula::SessionPoolConfig config;
+    config.username_ = "not_exists_user";
+    config.password_ = "nebula";
+    config.addrs_ = {kServerHost ":9669"};
+    config.spaceName_ = "session_pool_test";
+    config.maxSize_ = 2;
+    nebula::SessionPool pool(config);
+    ASSERT_FALSE(pool.init());
+  }
+
+  // user password error
+  {
+    nebula::SessionPoolConfig config;
+    config.username_ = "test_user";
+    config.password_ = "not_exists_password";
+    config.addrs_ = {kServerHost ":9669"};
+    config.spaceName_ = "session_pool_test";
+    config.maxSize_ = 2;
+    nebula::SessionPool pool(config);
+    ASSERT_FALSE(pool.init());
+  }
+}
+
+// case 2
+TEST_F(SessionPoolTest, Normal) {
+  // over connection limit
+  {
+    nebula::SessionPoolConfig config;
+    config.username_ = "root";
+    config.password_ = "nebula";
+    config.addrs_ = {kServerHost ":9669"};
+    config.spaceName_ = "session_pool_test";
+    config.maxSize_ = 2;
+    nebula::SessionPool pool(config);
+    ASSERT_TRUE(pool.init());
+    std::vector<std::thread> threads;
+    for (std::size_t i = 0; i < 5; ++i) {
+      threads.emplace_back([&pool]() {
+        auto resp = pool.execute("YIELD 1");
+        nebula::DataSet expected({"1"});
+        expected.emplace_back(nebula::Row({1}));
+        if (resp.errorCode == nebula::ErrorCode::SUCCEEDED) {
+          ASSERT_TRUE(verifyResultWithoutOrder(resp, expected));
+          ASSERT_EQ(*resp.spaceName, "session_pool_test");
+        } else {
+          ASSERT_EQ(*resp.errorMsg, "No idle session.");
+        }
+      });
+    }
+    for (auto& t : threads) {
+      t.join();
+    }
+  }
+}
+
+// case 3
+TEST_F(SessionPoolTest, SwitchSpace) {
+  nebula::SessionPoolConfig config;
+  config.username_ = "root";
+  config.password_ = "nebula";
+  config.addrs_ = {kServerHost ":9669"};
+  config.spaceName_ = "session_pool_test";
+  config.maxSize_ = 4;
+  nebula::SessionPool pool(config);
+  pool.init();
+  // switch space and auto switch back
+  {
```

Comment:
If you do not run  `USE session_pool_test2` and  `YIELD 1` in the same query, they might run via different sessions.

Simply do `execute("USE session_pool_test2; YIELD 1")` and check the space in the response. We do not need multithreaded here

Reply:
Good catch.